src/git-source-provider.ts

```diff
@@ -0,0 +1,292 @@
+                let lockPath = lockPaths[i];
+                try {
+                    await io.rmRF(lockPath);
+                }
+                catch (error) {
+                    core.debug(`Unable to delete '${lockPath}'. ${error.message}`);
+                }
+            }
+
+            // Clean
+            if (clean) {
+                if (!(await git.tryClean())) {
+                    core.debug(`The clean command failed. This might be caused by: 1) path too long, 2) permission issue, or 3) file in use. For futher investigation, manually run 'git clean -ffdx' on the directory '${repositoryPath}'.`);
+                    recreate = true;
+                }
+                else if (await git.tryReset()) {
+                    recreate = true;
+                }
+
+                if (recreate) {
+                    core.warning(`Unable to clean or reset the repository. The repository will be recreated instead.`);
+                }
+            }
+        }
+
+        // Recreate the directory
+        if (recreate) {
+            await io.rmRF(repositoryPath);
+            await io.mkdirP(repositoryPath);
+        }
+    }
+
+
+    // Initialize the repository
+    if (!fsHelper.directoryExistsSync(path.join(repositoryPath, '.git'))) {
+        await git.init();
+        await git.remoteAdd('origin', repositoryUrl);
+    }
+
+    // Disable automatic garbage collection
+    if (!(await git.tryDisableAutomaticGarbageCollection())) {
+        core.warning(`Unable to turn off git automatic garbage collection. The git fetch operation may trigger garbage collection and cause a delay.`);
+    }
+
+    // Remove possible previous extraheader
+    await removeGitConfig(git, authConfigKey);
+
+    // Add extraheader (auth)
+    let base64Credentials = Buffer.from(`x-access-token:${accessToken}`, 'utf8').toString('base64');
+    core.setSecret(base64Credentials);
+    let authConfigValue = `AUTHORIZATION: basic ${base64Credentials}`;
+    await git.config(authConfigKey, authConfigValue);
+
+    // LFS install
+    if (lfs) {
+        await git.lfsInstall();
+    }
+
+    // Fetch
+    let refSpec = getRefSpec(ref, commit);
+    await git.fetch(fetchDepth, refSpec);
+
+    // Checkout info
+    let checkoutInfo = await getCheckoutInfo(git, ref, commit);
+
+    // LFS fetch
+    // Explicit lfs-fetch to avoid slow checkout (fetches one lfs object at a time).
+    // Explicit lfs fetch will fetch lfs objects in parallel.
+    if (lfs) {
+        await git.lfsFetch(checkoutInfo.startPoint || checkoutInfo.ref);
+    }
+
+    // Checkout
+    await git.checkout(checkoutInfo.ref, checkoutInfo.startPoint);
+
+    // Dump some info about the checked out commit
+    await git.log1();
+
+    // Set intra-task state for cleanup
+    coreCommand.issueCommand('save-state', { name: 'repositoryPath' }, repositoryPath);
+    coreCommand.issueCommand('save-state', { name: 'configKey' }, authConfigKey);
+};
+
+export async function cleanup() {
```

Comment:
i need to revisit this and make sure its best effort (doesnt fail the step if the repo isnt there anymore or whatever)